docs/Overview.md

```diff
@@ -1,96 +1,246 @@
+The engine is equipped with one executors per language. An executor can translate a
+representation of a query in some language into another target language and delegate the execution to
+the executor for the target language.
+The translation process is carried out by compilers.
+Once a proper target language has been reached (possibly thanks to multiple compilation processes),
+the final representation of the query can be interpreted to produce results.
+During interpretation, the engine is responsible to provide the actual implementation of the data manipulation operations
+to the interpreter. Different engines can implement sources and transformations in different ways.
 
-There are several different concepts that make up the complete query engine.
+We now provide the high-level interfaces for the components described above:
 
-* Query - A query defines work to be performed on time series data and a result.
-    A query is represented as a directed acyclic graph (DAG).
-* Flux - Functional Language for defining a query to execute.
-* Parser - Parses a Flux script and produces a query.
-* Data Frame - A data frame is a matrix of time series data where one dimension is time and the other is series.
-* Query Node - A query node represents a single step in the DAG of a query.
-* Planner - The planner creates a plan of execution from a query.
-* Plan - A plan is also a DAG of node the explicitly state how a query will be performed.
-* Plan Node - A plan node represents a single step in the DAG of a plan.
-* Executor - The executor is responsible for processing a query plan.
-    The executor process data via data frames.
-* Storage - The Storage interface provides a mechanism for accessing the underlying data as data frames.
-* Capabilities - The Storage interface exposes its capabilities.
-    The planner uses the available capabilities to determine the best plan.
-* Hints - The Storage interface exposes hints about the data.
-    The planner uses the hints to determine the best plan.
-* Query Engine - Query Engine is the name given to the entire system being described in this document.
+The engine provides methods for registering sources and transformations, and for binding builtin values used for interpretation.
+Sources and transformations constitute the actual implementation of the data manipulation operations of some function values in the scope.
 
+```go
+type CreateSource = fn (ExecutionContext) Source
+type CreateTransformation = fn (ExecutionContext) Transformation
+type CompilerType interface {
+    From() Language
+    To() Language
+}
 
-Both a query and a plan are represented by a DAG and describe an operation that needs to be performed.
-The difference is that a plan in addition to describing what the operation is, also describes how that operation will be performed.
-In short, a query describes what the operation is and a plan describes how that operation will be carried out.
+type Engine interface {
```

Comment:
Every interface in here needs godoc comments.

I am going to wait to comment on the individual methods until there are the godoc comments. I would also prefer if this were reorganized so that the types like `Representation` were put higher up so that I could read it in the order of usage. It will make it easier to comment.

Reply:
I understand your point. I hadn't put godoc because every interface is introduced with its description on top. Anyways, I'll add it to the snippets.

The order in which the types are introduced is for better interweaving with the explanation outside of the snippet. If I change the order of the interfaces I would need to turn the explanation the other way around. It's something I'll do if the flow of the explanation is not clear as it is.